File: rm_ivalue/rm_ivalue/report/resource_allocation_status/resource_allocation_status.py
# Copyright (c) 2023, Yazan Hamdan and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, nowdate, add_days, date_diff, flt

def execute(filters=None):
    if not filters:
        filters = {}
        
    columns = get_columns()
    data = get_data(filters)
    
    chart_data = get_chart_data(data)
    
    return columns, data, None, chart_data

def get_columns():
    """Return columns for the report"""
    columns = [
        {
            "fieldname": "employee",
            "label": _("Employee ID"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120
        },
        {
            "fieldname": "employee_name",
            "label": _("Employee Name"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "department",
            "label": _("Department"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "project",
            "label": _("Project"),
            "fieldtype": "Link",
            "options": "Project",
            "width": 120
        },
        {
            "fieldname": "project_name",
            "label": _("Project Name"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "start_date",
            "label": _("Start Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "end_date",
            "label": _("End Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "allocation_percentage",
            "label": _("Allocation %"),
            "fieldtype": "Percent",
            "width": 100
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "remaining_days",
            "label": _("Remaining Days"),
            "fieldtype": "Int",
            "width": 100
        },
        {
            "fieldname": "estimated_cost",
            "label": _("Estimated Cost"),
            "fieldtype": "Currency",
            "width": 120
        }
    ]
    
    return columns

def get_data(filters):
    """Get data based on filters"""
    conditions = get_conditions(filters)
    
    # Query for Project Assignments
    data = frappe.db.sql("""
        SELECT 
            pa.employee,
            emp.employee_name,
            emp.department,
            pa.project,
            proj.project_name,
            pa.start_date,
            pa.end_date,
            pa.allocation_percentage,
            pa.status,
            pa.estimated_total_cost as estimated_cost,
            pa.name as assignment_id
        FROM 
            `tabProject Assignment` pa
        LEFT JOIN 
            `tabEmployee` emp ON pa.employee = emp.name
        LEFT JOIN 
            `tabProject` proj ON pa.project = proj.name
        WHERE 
            pa.docstatus = 1
            {conditions}
        ORDER BY 
            pa.start_date DESC
    """.format(conditions=conditions), as_dict=1)
    
    # Calculate remaining days for each assignment
    today = getdate(nowdate())
    for row in data:
        end_date = getdate(row.end_date)
        
        if today > end_date:
            row.remaining_days = 0
        else:
            row.remaining_days = date_diff(end_date, today)
    
    return data

def get_conditions(filters):
    """Build conditions for SQL query based on filters"""
    conditions = []
    
    if filters.get("employee"):
        conditions.append(" AND pa.employee = '{0}'".format(filters.get("employee")))
    
    if filters.get("project"):
        conditions.append(" AND pa.project = '{0}'".format(filters.get("project")))
    
    if filters.get("department"):
        conditions.append(" AND emp.department = '{0}'".format(filters.get("department")))
    
    if filters.get("status"):
        conditions.append(" AND pa.status = '{0}'".format(filters.get("status")))
    
    if filters.get("from_date"):
        conditions.append(" AND pa.start_date >= '{0}'".format(filters.get("from_date")))
    
    if filters.get("to_date"):
        conditions.append(" AND pa.end_date <= '{0}'".format(filters.get("to_date")))
    
    return " ".join(conditions)

def get_chart_data(data):
    """Generate chart data for the report"""
    if not data:
        return None
    
    # Prepare data for Project-wise allocation chart
    projects = {}
    for row in data:
        project = row.project_name or row.project
        if project in projects:
            projects[project] += flt(row.estimated_cost)
        else:
            projects[project] = flt(row.estimated_cost)
    
    project_labels = list(projects.keys())
    project_values = [projects[project] for project in project_labels]
    
    chart = {
        "type": "donut",
        "data": {
            "labels": project_labels,
            "datasets": [
                {
                    "values": project_values
                }
            ]
        },
        "colors": ["#5e64ff", "#7cd6fd", "#4caead", "#ff5858", "#ffa00a", "#26be8d", "#99cd9e", "#e6a9a9"],
        "height": 300,
    }
    
    return chart

File: rm_ivalue/rm_ivalue/report/resource_allocation_status/resource_allocation_status.js
// Copyright (c) 2023, Yazan Hamdan and contributors
// For license information, please see license.txt

frappe.query_reports["Resource Allocation Status"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), 3),
            "reqd": 0
        },
        {
            "fieldname": "employee",
            "label": __("Employee"),
            "fieldtype": "Link",
            "options": "Employee"
        },
        {
            "fieldname": "project",
            "label": __("Project"),
            "fieldtype": "Link",
            "options": "Project"
        },
        {
            "fieldname": "department",
            "label": __("Department"),
            "fieldtype": "Link",
            "options": "Department"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nActive\nCompleted\nCancelled"
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Add color to status field
        if (column.fieldname == "status") {
            if (data.status == "Active") {
                value = "<span style='color:green'>" + value + "</span>";
            } else if (data.status == "Completed") {
                value = "<span style='color:blue'>" + value + "</span>";
            } else if (data.status == "Cancelled") {
                value = "<span style='color:red'>" + value + "</span>";
            }
        }
        
        // Highlight rows with low remaining days
        if (column.fieldname == "remaining_days" && data.status == "Active") {
            if (data.remaining_days <= 7) {
                value = "<span style='color:red; font-weight:bold'>" + value + "</span>";
            } else if (data.remaining_days <= 14) {
                value = "<span style='color:orange; font-weight:bold'>" + value + "</span>";
            }
        }
        
        // Add progress bar for allocation percentage
        if (column.fieldname == "allocation_percentage") {
            let percentage = data.allocation_percentage || 0;
            let progressBar = `
                <div class="progress" style="margin-bottom: 0; height: 12px;">
                    <div class="progress-bar" role="progressbar" 
                        aria-valuenow="${percentage}" 
                        aria-valuemin="0" 
                        aria-valuemax="100" 
                        style="width: ${percentage}%; background-color: ${percentage > 80 ? '#ff5858' : '#5e64ff'};">
                        ${percentage}%
                    </div>
                </div>
            `;
            value = progressBar;
        }
        
        return value;
    }
};
